### DISCUSSION TITLE
Using with older version of laravel

### DISCUSSION THREAD

**Discussion by Dante983** [adalessa/laravel.nvim#102] [Ideas]
I was just wondering is it possible to set this to use with older versions of laravel?
As at one project i am using laravel 5.4 atm (there is a plan to upgrade this), and there the plugin is just not working, for artisna commadn i get bunch of error/warnings and that's it.

Maybe I missed something but a feature like where you could set the version of larvel that you want would be awesome.

<img width="1799" alt="image" src="https://github.com/user-attachments/assets/78bf2cb3-c8f4-4800-81e9-3f5af9264ecd">

**Comment by adalessa** [comment]
Possible is for sure, assuming you are on development branch soon to be merge, the solution could be override the services which uses non compatible ones. 
For this assume the error is on the getting the config, so it will be in the confiigs repository, there could be change.
That could be a different plugin that offers the providers and the user can just register it.

**Comment by mattyforth** [comment]
I'm also struggling with this, occurs when the PHP version in your path is too high for the codebase. 

I have multiple PHP versions installed on my machine, managed with [Laravel Valet](https://laravel.com/docs/12.x/valet). One solution / workaround I'm trying to get working is detecting what version to use using Valet and then replacing the php executable in the `environments` config to be the full qualified path of the PHP versions.

Something like this: 
```lua
  opts = function()
    local opts = require('laravel.options.default')
    local environments = opts.environments
    local definitions = environments.definitions

    for key, definition in pairs(definitions) do
      if definition.name == 'local' then
        definition.condition.executable = { get_php_path() }
      end
    end

    return opts
  end,
  config = true,
```
This results in the following:
```lua
{
        condition = {
          executable = { "/opt/homebrew/opt/php@8.3/bin/php" }
        },
        name = "local"
}
```
However that doesn't seem to have any affect on the plugin and it still doesn't use the updated PHP path. I have also tried using `config = function` and passing the options to `.setup(opts)`, however that doesn't work either.

I've also tried setting up an extra definition into the environments table for Valet similar to the Herd definition, however I'm facing similar issues in that the config just doesn't seem to respect the changes I make. 
```lua
    local opts = require('laravel.options.default')
    local environments = opts.environments
    local definitions = environments.definitions

    environments.default = 'valet'

    table.insert(definitions, {
      name = "valet",
      condition = {
        executable = { "valet" },
      },
      commands = {
        valet = { "valet" },
        {
          commands = { "php", "composer" },
          prefix = { "valet" },
        },
      },
    })

    return opts
```

It might be that I need to add a provider that adds Valet support instead? I might be on the complete wrong path here so any help would be greatly appreciated!

**Reply by adalessa** [reply]
Hi @mattyforth
What you changed first was just the condition to see if the environment should be use. not the command to use.
So because of that has not the desire effect.
I am not sure how valet work now a days, if you run "valet php" run the current version of php, because that is what would be achieve with the prefix idea, which works for herd but don't know if it works for valet.

Before I have not thought about that need of custom path, the current is more to add prefix like with sail or docker compose.

Please confirm if valet works in that way. Thanks

**Reply by mattyforth** [reply]
Ah that makes complete sense, apologies I didn't think it was working like that. 

Yes so when you use `valet php`, it'll look up the php version to use for your project and use the correct version, it's pretty much identical to herd in this regard. 

Thanks for getting back to me, really appreciate it!

**Reply by mattyforth** [reply]
I've managed to get this working after some looking around in the source code, all I needed to change was `auto_discover` to `false`. When `auto_discover` is on, it will take priority over what you have set for your default, so the first executable it finds (in this case the default PHP version in your path) will be used. So for anyone that wants to use Laravel valet, you can use this configuration.

```lua
  opts = function()
    local opts = require('laravel.options.default')
    local environments = opts.environments
    local definitions = environments.definitions

    environments.default = 'valet'
    environments.auto_discover = false

    table.insert(definitions, {
      name = "valet",
      condition = {
        executable = { "valet" },
      },
      commands = {
        valet = { "valet" },
        {
          commands = { "php", "composer" },
          prefix = { "valet" },
        },
      },
    })

    return opts
  end,
```

Thanks @adalessa for your help the other day! I can look into making a PR to add this into the default definitions if that's something you are interested in?

**Reply by adalessa** [reply]
If you want for now I will take the PR, but I am working in a new way for 4.x version which will make that step unecessary.
